File: app/src/main/java/workshop/akbolatss/tagsnews/screen/news/NewsAdapter.java
package workshop.akbolatss.tagsnews.screen.news;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.squareup.picasso.Picasso;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import workshop.akbolatss.tagsnews.R;
import workshop.akbolatss.tagsnews.screen.news.model.Article;

/**
 * Created by AkbolatSS on 08.08.2017.
 */

public class NewsAdapter extends RecyclerView.Adapter<NewsAdapter.NewsHolder> {


    private LayoutInflater mLayoutInflater;
    private List<Article> mNewsList;

    public NewsAdapter(LayoutInflater mLayoutInflater) {
        this.mLayoutInflater = mLayoutInflater;
        mNewsList = new ArrayList<>();
    }

    @Override
    public NewsHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = mLayoutInflater.inflate(R.layout.rv_news_item, parent, false);
        return new NewsHolder(view);
    }

    @Override
    public void onBindViewHolder(NewsHolder holder, int position) {
        holder.bind(mNewsList.get(position));
    }

    @Override
    public int getItemCount() {
        return mNewsList.size();
    }

    public void onAddArticles(List<Article> articles) {
        mNewsList.addAll(articles);
        notifyDataSetChanged();
    }

    public class NewsHolder extends RecyclerView.ViewHolder {

        private Context mContext;
        @BindView(R.id.imgView)
        ImageView mImage;
        @BindView(R.id.tvTitle)
        TextView mTitle;
        @BindView(R.id.tvDescription)
        TextView mDescription;
        @BindView(R.id.tvTimestamp)
        TextView mTimestamp;

        public NewsHolder(View itemView) {
            super(itemView);
            mContext = itemView.getContext();
            ButterKnife.bind(this, itemView);
        }

        public void bind(Article article) {
            mTitle.setText(article.getTitle());
            mDescription.setText(article.getDescription());

            mTimestamp.setText(article.getTimestamp());

            Picasso.with(mContext).load(article.getImageUrl())
                    .into(mImage);
        }
    }
}

File: app/src/main/java/workshop/akbolatss/tagsnews/screen/news/NewsAdapter.kt
package workshop.akbolatss.tagsnews.screen.news

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.squareup.picasso.Picasso
import com.stfalcon.frescoimageviewer.ImageViewer
import kotlinx.android.synthetic.main.rv_news_item_small.view.*
import me.toptas.rssconverter.RssItem
import workshop.akbolatss.tagsnews.R
import workshop.akbolatss.tagsnews.util.Constants.ITEM_VIEW_MAIN
import workshop.akbolatss.tagsnews.util.UtilityMethods
import java.util.*

/**
 * Custom RecyclerView.Adapter for News items
 */
class NewsAdapter(private val mClickListener: NewsListener) : RecyclerView.Adapter<NewsAdapter.NewsHolder>() {

    /**
     * List of RSS feeds
     */
    private val mNewsList: MutableList<RssItem> = ArrayList()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): NewsHolder {
        val mLayoutInflater = LayoutInflater.from(parent.context)
        val view: View = mLayoutInflater.inflate(R.layout.rv_news_item_small, parent, false)
        return NewsHolder(view)
    }

    override fun onBindViewHolder(holder: NewsHolder, position: Int) {
        val rssItem = mNewsList[position]
        holder.bind(rssItem, mClickListener)
    }

    override fun getItemCount(): Int {
        return mNewsList.size
    }

    override fun getItemViewType(position: Int): Int {
        return ITEM_VIEW_MAIN
    }

    /**
     * Load list of items to this adapter
     */
    fun onAddItems(rssItems: List<RssItem>?) {
        if (rssItems != null) {
            mNewsList.clear()
            mNewsList.addAll(rssItems)
            notifyDataSetChanged()
        }
    }

    interface NewsListener {
        fun onItemClick(rssItem: RssItem)
    }

    /**
     * Custom View Holder where View instantiates
     */
    inner class NewsHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        fun bind(rssItem: RssItem, listener: NewsListener) {

            itemView.tvTitle.text = rssItem.title
            if (rssItem.publishDate != null && rssItem.publishDate.isNotEmpty())
                itemView.tvTimestamp.text = UtilityMethods.convertTime(rssItem.publishDate)

            itemView.tvDescription.text = rssItem.description

            if (rssItem.image != null) {
                itemView.imgImage.visibility = View.VISIBLE
                Picasso.with(itemView.context)
                        .load(rssItem.image)
                        .placeholder(R.drawable.placeholder)
                        .into(itemView.imgImage)
            } else {
                itemView.imgImage.visibility = View.GONE
            }

            itemView.flItem.setOnClickListener {
                listener.onItemClick(rssItem)
            }
            itemView.flImage.setOnClickListener {
                ImageViewer.Builder(itemView.context, arrayListOf(rssItem.image))
                        .setStartPosition(0)
                        .show()
            }
        }
    }
}
